torchaudio/functional.py

```diff
@@ -305,3 +305,176 @@ def mu_law_expanding(x_mu, qc):
     x = ((x_mu) / mu) * 2 - 1.
     x = torch.sign(x) * (torch.exp(torch.abs(x) * torch.log1p(mu)) - 1.) / mu
     return x
+
+
+def stft(waveforms, fft_length, hop_length=None, win_length=None, window=None,
```

Comment:
My understanding is this function takes waveforms of size (..., signal_len) and does stft on the last dimension to return size (..., freq, time, 2).

Is there a way to incorporate this into [spectrogram](https://github.com/pytorch/audio/blob/master/torchaudio/functional.py#L109)? I think it would be preferable to have a single function for stft

Reply:
It'd be definitely possible and we can also remove `_stft()` then. But I understood this PR to be not changing the previous implementation and behavior, hence not sure about it. (It should definitely happen after the freezing anyway)

Reply:
What's the advantage of this over torch.stft? We're in the process of adding istft which will be build around being a sane inverse of torch.stft. Having torchaudio.stft and torch.stft will be confusing.

Reply:
AFAIR the issue was that we didn't like the default names/values/behaviors of `torch.stft`. I think this also boils down to to what extent we'd like these functionals to be less hacky in the phase 1.

Reply:
@keunwoochoi - we'll need to revisit the discussion around the default names, values and behavior. We cannot add a torchaudio.stft, it'll be very confusing and hacks around an issue that's related to the core of pytorch and we have the power resolve over there.

Reply:
@keunwoochoi maybe initially, but at this point the default values are the same as `torch.stft`. I think it was more about the names and `torchaudio.stft` being multichannel.

Reply:
I see there is https://github.com/keunwoochoi/torchaudio-contrib/issues/36 in torchaudio-contrib around this. The main thing I can tell from your code is that you want batch support for the input (waveforms) and the hann_window as a standard if none is specified. You're also implying the preference of input to be named "waveforms" instead of using [input in torch core](https://pytorch.org/docs/master/torch.html#torch.stft).

You're also implying the standard of having "channel" first, which is something @jamarshon is looking into as the framework-wide standard. We talked about this today and it seems the most sane, since it's ML specific (and this is an ML library) and since its consistent with the standards used in common operations [such as conv1d](https://pytorch.org/docs/master/nn.html#torch.nn.Conv1d).

Reply:
cc @ksanjeevan

Reply:
We talked about this separately again today and the main thing this wrapper does is add batching support.

Batching is a big topic overall. I'd prefer if we would add batching support in a separate PR and, if we add it, do so globally, since none of the other transforms have batching support either.

Once you add batching support user will probably want to have support for masks etc., so then we should think it about it more principled or maybe even wait for something like NestedTensor.

This also doesn't block people from using batching if they're aware of the reshape trick.

So, could I please ask you to remove this from the PR and then send again later after we talked more about batching (and after the standardization)?

Reply:
cc @ksanjeevan